File: solutions/0003-longest-substring-without-repeating-characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char,int>m;
        int left = 0 ;
        int longest = 0;
        for(int right=0;right<s.size();right++){
            m[s[right]]++;

            while(left<=right and m[s[right]]!=1){
                m[s[left]]--;
                left++;
            }
            longest = max(longest,right-left+1);
        }
        return longest;
    }
};

File: solutions/0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans = 0
        left = 0
        m = {}
        for right in range(len(s)):
            if(s[right] not in m):
                m[s[right]]=1
            else:
                m[s[right]]+=1
            # check if invalid state
            while(m[s[right]]>1):
                m[s[left]]-=1
                left+=1
            # check the window
            ans = max(right-left+1,ans)
        return ans
            



        
